src/Microsoft.Health.Fhir.Api/Features/Context/FhirRequestContextMiddleware.cs

```diff
@@ -40,7 +41,10 @@ public async Task Invoke(HttpContext context, IFhirRequestContextAccessor fhirRe
                 request.Path,
                 request.QueryString);
 
-            string correlationId = correlationIdProvider.Invoke();
+            string requestId = correlationIdProvider.Invoke();
+            string correlationId = null;
+            var key = context.Request.Headers.Keys.FirstOrDefault(h => h.Equals(KnownHeaders.CorrelationId, StringComparison.OrdinalIgnoreCase));
```

Comment:
I interpreted the spec differently...

`X-Request-Id`: A unique id to for the request/response assigned by either client or server. **Request: assigned by the client. Response: assigned by the server**
`X-Correlation-Id`: A client assigned request id echoed back in the response
(https://www.hl7.org/fhir/http.html#custom)

So example:
- Client sends header `X-Request-Id`: ABC
- Server response: `X-Correlation-Id`: ABC, `X-Request-Id`: 123

The client now has a mapping `ABC : 123` between client / server request logs.

Thoughts? Differences?

Reply:
One day I'll learn to read every single line in spec...
Didn't notice clarification under table.

From what I read (without noticing how it actually should implemented) I thought what request id is just something we return always and unique, and correlation is something user can specify and we can pass around. Sound weird for me to constantly have header mapping between ReadResponse->GetCorrelationId-> Shove it into new Request under RequestId, but if that was spec wants...

---
In reply to: [589812549](https://github.com/microsoft/fhir-server/pull/1728#discussion_r589812549) [](ancestors = 589812549)